### PR TITLE
Migrate buildSearchPostFilterClause to use Squirrel

### PR DESCRIPTION
**Pull request by superkkt**
#### Summary

This PR migrates buildSearchPostFilterClause to use Squirrel.

#### Ticket Link

Fixes https://github.com/mattermost/mattermost-server/issues/18428

#### Release Note

```release-note
NONE
```

### CODE REVIEW

File: store/sqlstore/post_store.go [hunk 1]
```diff
@@ -1684,41 +1684,36 @@ func (s *SqlPostStore) buildSearchUserFilterClause(users []string, paramPrefix s
...
-	excludedUserClause, queryParams := s.buildSearchUserFilterClause(excludedUsers, "ExcludedUser", true, queryParams, userByUsername)
-	filterQuery = strings.Replace(filterQuery, "EXCLUDED_USER_FILTER", excludedUserClause, 1)
+	// Sub-query builder.
+	sb := sq.Select("Id").From("Users, TeamMembers").Where("TeamMembers.TeamId = :TeamId").Where("Users.Id = TeamMembers.UserId")
+	sb, queryParams = s.buildSearchUserFilterClause(fromUsers, "FromUser", false, queryParams, userByUsername, sb)
+	sb, queryParams = s.buildSearchUserFilterClause(excludedUsers, "ExcludedUser", true, queryParams, userByUsername, sb)
+	subQuery, _, err := sb.ToSql()
+	// This should not fail.
+	if err != nil {
+		panic(err)
```
Comment (larkox) [inline store/sqlstore/post_store.go:1709]: I am not a fan of panics, but I will leave the decision on what to do in this situation to @agnivade .

**Review by larkox — APPROVED** [review]
LGTM, just one small concern.

File: store/sqlstore/post_store.go [hunk 2]
```diff
@@ -1684,41 +1684,36 @@ func (s *SqlPostStore) buildSearchUserFilterClause(users []string, paramPrefix s
 	clause := strings.Join(clauseSlice, ", ")
```
Comment (agnivade) [inline store/sqlstore/post_store.go:1684]: This needs to be changed to the actual args as well. This still uses the old style of ":paramName" to bind arguments.

Overall, a lot of this code needs to be rewritten to use squirrel completely and not rely string operations. Which means `buildSearchPostFilterClause`, `buildCreateDateFilterClause` need to be converted as well.

We should move out of `map[string]interface{}` completely.

Sorry, I should have clarified this earlier in the issue itself. This is a relatively big task. Let's take our time and do it properly.

File: store/sqlstore/post_store.go (hunk 1 above)
Comment (agnivade) [reply store/sqlstore/post_store.go:1709]: Yes, please bubble the error upwards and handle it.

**Review by agnivade — CHANGES_REQUESTED** [review]
Looks like a great start @superkkt ! However, we need some more work here.

**Comment by superkkt** [comment]
> Looks like a great start @superkkt ! However, we need some more work here.

@agnivade Okay. I'm working on it now.

**Comment by superkkt** [comment]
@agnivade Please review again and let me know if we need more changes.

File: store/sqlstore/post_store.go [hunk 3]
```diff
@@ -1583,154 +1583,119 @@ var specialSearchChar = []string{
...
 	// handle after: before: on: filters
 	if params.OnDate != "" {
 		onDateStart, onDateEnd := params.GetOnDateMillis()
-		queryParams["OnDateStart"] = strconv.FormatInt(onDateStart, 10)
-		queryParams["OnDateEnd"] = strconv.FormatInt(onDateEnd, 10)
-
 		// between `on date` start of day and end of day
-		builder = builder.Where("CreateAt BETWEEN :OnDateStart AND :OnDateEnd")
-		return builder, queryParams
+		builder = builder.Where("CreateAt BETWEEN ? AND ?", strconv.FormatInt(onDateStart, 10), strconv.FormatInt(onDateEnd, 10))
```
[the remark this answers is not in the document]
Comment (superkkt) [reply store/sqlstore/post_store.go]: @agnivade You are right. It works. I just copied previous codes unintentionally.

**Comment by superkkt** [comment]
@agnivade Could you please review again?

**Review by agnivade — APPROVED** [review]
Nicely done!

**Pull request merged**